### ISSUE TITLE
Custom stop loss function is not working on dry-run mode (?)

### ISSUE THREAD

**Issue by win-lobo** [freqtrade/freqtrade#9110]
## Describe your environment
  * Operating system: DietPi v8.19.1
  * Python Version: 3.11.2
  * CCXT version: 3.1.44
  * Freqtrade Version: 2023.6.dev-e16c433c
  
## Your question

I have been testing a strategy in dry-run and realized that the custom stop loss is working when the trade is going towards profit direction, however, the custom stop loss is also trailing down when the trade turns back. For example: The trade in the screenshot below was in 34.34% profit at some point, so the stoploss should be set to 17%, which the stoploss distance is showing correctly (image 1). But when the trade was down to 22.30% (image 2), the stoploss distance pointed the stoploss at 14.18%.
![Screenshot from 2023-08-26 18-52-39](https://github.com/freqtrade/freqtrade/assets/33344125/8e2beae9-88aa-4335-8f67-1c7c15b4f462)  ![Screenshot from 2023-08-26 18-53-17](https://github.com/freqtrade/freqtrade/assets/33344125/4e8ec47e-337f-45b1-96ce-263076663359)

According to the documentation the [stoploss price can only ever move upwards](https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-stoploss), which is clearly not happening here.
- Is there a chance that the custom stoploss is not working on dry-run but only on live?
- Will the helper function stoploss_from_open() prevent this? if not, what else can be useful for trailing the stop correctly even in dry-run?

My custom_stoploss function is below for reference.

```python
use_custom_stoploss = True

    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:

        # evaluate highest to lowest, so that highest possible stop is used
        if current_profit < 0.05:
            return -1 # return a value bigger than the initial stoploss to keep using the initial stoploss

        # After reaching the desired offset, allow the stoploss to trail by half the profit
        desired_stoploss = current_profit / 2

        # Use a minimum of 2.5% and a maximum of 5%
        return max(min(desired_stoploss, current_profit), 0.025)

        # return maximum stoploss value, keeping current stoploss price unchanged
        #return 1

```

**Comment by stash86** [comment]
If you look clearly at the pictures, it clearly shows that your stoploss is at 0.4885 (-19.04%) for both cases, so it never move down

**Comment by win-lobo** [comment]
I see that now - but still it seems confusing to me, because if my stoploss is set on 0.4885. that's approx. 10% profit from open rate. The highest current profit was 34%, the stoploss should be set at 17% at least. Could you explain me that, please?

**Comment by stash86** [comment]
before that, can you give me what is your expected stoploss rate (not percentage) when the profit is 0%, 5%, 10%, 15%, 20%, 25%, 30% and 35%?

**Comment by stash86** [comment]
hint
`The method must return a stoploss value (float / number) as a percentage of the current price. E.g. If the current_rate is 200 USD, then returning 0.02 will set the stoploss price 2% lower, at 196 USD.`

**Comment by win-lobo** [comment]
0% -> 0.03 (initial stoploss)
5%
10%
15%
20%
25%30% and 35%?

**Comment by win-lobo** [comment]
Expected stoploss rate:
0% -> -0.03 (initial stoploss)
5% -> -0.03 (initial stoploss)
10% -> 0.05
15% -> 0.075
20% -> 0.1
25% -> 0.125
30% -> 0.15
35%  ->  0.175

**Comment by stash86** [comment]
I'm not asking for the percentage. I'm asking the actual stoploss rate.

**Comment by win-lobo** [comment]
Sorry, I think I understand your point now, I was misunderstanding the concept of stoploss rate.

Expected stoploss rate:
0% -> -0.4254 (initial stoploss rate)
5% -> -0.4254 (initial stoploss rate)
10% -> 0.460425
15% -> 0.075
20% -> 0.48235
25% -> 0.4933125
30% -> 0.504275
35% -> 0.5152375

I now get it. So, the helper function stoploss_from_open() is helpful here if I want to get the stoploss rate from open rate, right?

**Comment by xmatthias** [comment]
actually, i think your stoploss is set correctly / as expected.

the misunderstanding is a mathematical one.
to make a simplified example:
100 + 25% = 125
but 125 - 25% != 100 - but ~93%.


As such, your (most likely) math of 22.3% - 8.12% is not really working out (calculating with percentages is not as easy as aggregating 2 values together).

**Comment by stash86** [comment]
Yes, the bot calculated it right. I just want to test their knowledge, because I want to know how they calculate the stoploss rate.

this is how the stoploss is calculated, and as you can see, matched the bot
https://docs.google.com/spreadsheets/d/14r9mapno4okiuM5U_iqN26MNsW_b748FdZB4TlnnR_A/edit?usp=sharing

**Comment by win-lobo** [comment]
Thank you guys for your help. I understand now that the stoploss is calculated from the highest current_rate (not from open rate) so it will be something like you guys are pointing out and the bot is working as it should be. I will need to change my `custom_stoploss()` function to return a stoploss rate based on the open rate instead. 
I guess I will need to look into this `stoploss_from_open()` helper. Thanks again!
